File: Data-Structures/Binary-Search-Tree/bst.js
class Node {
  constructor(value) {
    this.value = value;
    this.right = null;
    this.left = null;
  }
}

class Tree {
  constructor() {
    this.root = null;
  }
  //invert binary tree
  invert(){
    this,this._invert(this.root);
  }
  _invert(node){
    if(!node){
      return;
    }
    //swap left and right subtrees
    const temp = node.left;
    node.left=node.right;
    node.right = temp;
    this._invert(node.left)
    this._invert(node.right)
  }

  //insert
  insert(currentNode, data) {
    if (!currentNode) {
      return new Node(data);
    }

    if (data < currentNode.value) {
      currentNode.left = this.insert(currentNode.left, data);
    } else if (data > currentNode.value) {
      currentNode.right = this.insert(currentNode.right, data);
    }

    return currentNode;
  }
  insertNode(data) {
    this.root = this.insert(this.root, data);
  }
  //delete
  deleteNode(data) {
    this.root = this._delete(this.root, data);
  }

  _delete(node, data) {
    if (!node) {
      return null;
    }

    if (data === node.value) {
      // Case 1: Node with only one child or no child
      if (!node.left) {
        return node.right;
      } else if (!node.right) {
        return node.left;
      }

      // Case 2: Node with two children
      node.value = this._findMin(node.right).value;
      node.right = this._delete(node.right, node.value);
    } else if (data < node.value) {
      node.left = this._delete(node.left, data);
    } else if (data > node.value) {
      node.right = this._delete(node.right, data);
    }

    return node;
  }

  _findMin(node) {
    while (node.left) {
      node = node.left;
    }
    return node;
  }

  //search
  find(data) {
    return this._find(this.root, data);
  }
  _find(node, data) {
    if (node === null) {
      return false;
    }
    if (node.value === data) {
      return true;
    } else if (data > node.value) {
      return this._find(node.right, data);
    } else if (data < node.value) {
      return this._find(node.left, data);
    }
  }
  //traversal
  //DFS
  preOrder() {
    let result = [];

    if (!this.root) {
      return;
    }

    function traverse(currentNode) {
      result.push(currentNode.value);
      if (currentNode.left) {
        traverse(currentNode.left);
      }
      if (currentNode.right) {
        traverse(currentNode.right);
      }
    }
    traverse(this.root);

    return result;
  }
  postOrder() {
    let result = [];

    if (!this.root) {
      return;
    }

    function traverse(currentNode) {
      if (currentNode.left) {
        traverse(currentNode.left);
      }
      if (currentNode.right) {
        traverse(currentNode.right);
      }
      result.push(currentNode.value);
    }

    traverse(this.root);
    return result;
  }

  inOrder() {
    let result = [];
    if (!this.root) {
      return;
    }
    function traverse(currentNode) {
      if (currentNode.left) {
        traverse(currentNode.left);
      }
      result.push(currentNode.value);
      if (currentNode.right) {
        traverse(currentNode.right);
      }
    }

    traverse(this.root);
    return result;
  }
  //BFS

  BFS() {
    if (!this.root) {
      return [];
    }
    let currentNode = this.root;
    //array to hold nodes
    let queue = [];
    //array to push node.value
    let result = [];
    //initially push root
    queue.push(currentNode);
    while (queue.length) {
      //set current node to first value of queue

      currentNode = queue.shift();
      //push current node to result array
      result.push(currentNode.value);

      if (currentNode.left) {
        //if a left node exists push it to the queue
        queue.push(currentNode.left);
      }
      if (currentNode.right) {
        //if a right node exists push it to the queue
        queue.push(currentNode.right);
      }
    }
    return result;
  }

  //finding the minimum value of the tree
  min() {
    return this._min(this.root);
  }
  _min(currentNode) {
    if (!currentNode.left) {
      return currentNode.value;
    } else {
      return this._min(currentNode.left);
    }
  }
  //finding the maximum value of the tree
  max() {
    return this._max(this.root);
  }
  _max(currentNode) {
    if (!currentNode.right) {
      return currentNode.value;
    } else {
      return this._max(currentNode.right);
    }
  }
  //height
  height() {
    return this._height(this.root);
  }
  _height(currentNode) {
    if (!currentNode) {
      return 0;
    }
    let leftHeight = this._height(currentNode.left);
    let rightHeight = this._height(currentNode.right);

    return Math.max(leftHeight, rightHeight) + 1;
  }
}

File: Data-Structures/Binary-Search-Tree/bst.py
class Node:
    def __init__(self,value):
        self.value = value
        self.left = None
        self.right = None

class Tree:
    def __init__(self) :
        self.root = None

    #insert
    def insert(self,value):
        self.root = self._insert(self.root,value)

    def _insert(self,currentNode,value):
        if not currentNode:
            return Node(value)

        if value < currentNode.value:
            currentNode.left = self._insert(currentNode.left,value)
        elif value > currentNode.value:
            currentNode.right = self._insert(currentNode.right,value)
        return currentNode            
    #delete
    def delete(self, value):
        self.root = self._delete(self.root, value)

    def _delete(self, node, value):
        if not node:
            return None
        if value == node.value:
            # case 1: node with only one child or no child
            if not node.left:
                return node.right
            elif not node.right:
                return node.left

            # case 2: node with two children
            node.value = self._find_min(node.right).value
            node.right = self._delete(node.right, node.value)
        elif value < node.value:
            node.left = self._delete(node.left, value)
        elif value > node.value:
            node.right = self._delete(node.right, value)

        return node

    def _find_min(self, node):
        while node.left:
            node = node.left
        return node


    #max
    def max(self):
        return self._max(self.root)
    
    def _max(self,currentNode):
        if not currentNode.right:
            return currentNode.value
        else:
            return self._max(currentNode.right)
    #min
    def min(self):
        return self._min(self.root)

    def _min(self,currentNode):
        if not currentNode.left:
            return currentNode.value
        else:
            return self._min(currentNode.left)
    #find value
    def find(self,value):
        return self._find(self.root,value)

    def _find(self,currentNode,value):
        if not currentNode:
            return False
        if currentNode.value == value:
            return True
        if value > currentNode.value:
            return self._find(currentNode.right,value)
        elif value < currentNode.value:
            return self._find(currentNode.left,value)
        return False    
    #Traversal
    #BFS
    def BFS(self):
        if not self.root:
            return []
        queue = []
        result = []
        
        queue.append(self.root)

        while len(queue):
            currentNode = queue.pop(0)
            result.append(currentNode.value)

            if currentNode.left:
                queue.append(currentNode.left)
            if currentNode.right:
                queue.append(currentNode.right)    
        return result    
    #DFS
    #pre_order
    def pre_order(self):
        result = []
        if not self.root:
            return
    
        def walk(currentNode):
            
            result.append(currentNode.value)
            if currentNode.left:
                walk(currentNode.left)
            if currentNode.right:
                walk(currentNode.right)    

        walk(self.root)
        return result    
    #post_order
    def post_order(self):
        if not self.root:
            return
        result = []

        def walk(currentNode):
            if currentNode.left:
                walk(currentNode.left)
            result.append(currentNode.value)
            if currentNode.right:
                walk(currentNode.right)
        walk(self.root)
        return result        

    #in_order
    def in_order(self):
        if not self.root:
            return
        result = []

        def walk(currentNode):
            if currentNode.left:
                walk(currentNode.left)
            if currentNode.right:
                walk(currentNode.right)
            result.append(currentNode.value)
        walk(self.root)
        return result        

    #height
    def height(self):
        return self._height(self.root)
    def _height(self,currentNode):
        if not currentNode:
            return 0
        
        leftHeight = self._height(currentNode.left)
        rightHeight = self._height(currentNode.right)
        
        height = max(leftHeight,rightHeight) + 1
        return height



